### ISSUE TITLE
Wrong recommendation for `needless_range_loop` if loop variable is used in closure

### ISSUE THREAD

**Issue by vbrandl** [rust-lang/rust-clippy#2542]
Clippy falsely warns about a needless range loop, if the loop variable is (aside from indexing) only used inside a closure. The following example triggers the false positive:

```Rust
fn option() -> Option<bool> {
    unimplemented!()
}

fn error(idx: usize) -> () {
    unimplemented!()
}

fn test(len: usize) {
    let mut arr = Vec::with_capacity(len);
    for idx in 0..len {
        arr[idx] = option().ok_or_else(|| error(idx)).unwrap();
    }
}
```

This could be written as

```Rust
fn test(len: usize) {
    let mut arr = Vec::with_capacity(len);
    for (idx, val) in arr.iter_mut().take(len).enumerate() {
        *val = option().ok_or_else(|| error(idx)).unwrap();
    }
}
```

But clippy says to use `for <item> in arr.iter_mut().take(len) {`

Tested with `rustc 1.25.0-nightly (3ec5a99aa 2018-02-14)` and clippy `0.0.186`.
